rest-api/dao/participant_dao.py

```diff
@@ -114,25 +114,49 @@ def _do_update(self, session, obj, existing_obj):
                             else None)
       need_new_summary = True
 
-    if obj.organizationId or obj.siteId or obj.hpoId:
-      site, organization, awardee = self.get_pairing_level(obj)
-      obj.organizationId = organization
-      obj.siteId = site
-      obj.hpoId = awardee
-      if awardee:
-        # get provider link for hpo_id (awardee)
-        obj.providerLink = make_primary_provider_link_for_id(awardee)
+    update_pairing = True
+    if obj.siteId is None and obj.organizationId is None and obj.hpoId is None and \
+      obj.providerLink == 'null':
+      # Prevent unpairing if /PUT is sent with no pairing levels.
+      update_pairing = False
+
+    if update_pairing == True:
+      # site,org,or awardee is sent in request: Get relationships and try to set provider link.
+      if obj.organizationId or obj.siteId or (obj.hpoId >= 0) and (obj.providerLink ==
+                                                            existing_obj.providerLink or
+                                                            obj.providerLink == 'null'):
+        site, organization, awardee = self.get_pairing_level(obj)
+        obj.organizationId = organization
+        obj.siteId = site
+        obj.hpoId = awardee
+        if awardee is not None and (obj.hpoId != existing_obj.hpoId):
+          # get provider link for hpo_id (awardee)
+          obj.providerLink = make_primary_provider_link_for_id(awardee)
 
-      need_new_summary = True
+        need_new_summary = True
 
-    # If the provider link changes, update the HPO ID on the participant and its summary.
-    if obj.hpoId is None:
+    if update_pairing == True:
+      # If the provider link changes, update the HPO ID on the participant and its summary.
+      if obj.hpoId is None:
+        obj.hpoId = existing_obj.hpoId
+      if obj.providerLink != existing_obj.providerLink:
+        # check/set hpoId (covers the chance of sending only provider link and not entering logic
+        # above with hpoId.
+        new_hpo_id = self._get_hpo_id(obj)
+        if new_hpo_id != existing_obj.hpoId:
+          obj.hpoId = new_hpo_id
+          if obj.hpoId == UNSET_HPO_ID:
```

Comment:
So I think actually this and the next two lines are unnecessary, because obj.siteId and obj.organizationId won't have been set to anything above in this case. Is that right?

If they ever *were* initialized to the existing site ID and organization ID values, we'd need to clear them out when the provider link changes... so I'd say at least this if statement should be removed.

Please add a test for the case where you're paired with a site and the a provider link comes in for another awardee.

Reply:
done.